README.markdown
Happy Phone Number [![Build Status](https://travis-ci.org/lkdjiin/happy_phone_number.png)](https://travis-ci.org/lkdjiin/happy_phone_number)
================

Description
----------------------

*Happy Phone Number* provides easy methods to format phone
numbers from your Rails models. Virtually any country specific
format or convention could be supported, including national and international
format. For the not (yet) supported countries, a general method
using a simple mask is also provided.

Overview
-----------------------

### Prerequisite

  * Phone number **must be save** with digits only. So, without spaces, hyphens,
    or dots. Just digits.

### Demonstration

Let's say you have a `Contact` model with a `phone` attribute.

To display the phone number you could do this:

    <%= @contact.phone %>
    #=> "0123456789"

Using *Happy Phone Number* you could do that:

    <%= @contact.happy_phone(:fr) %>
    #=> "01 23 45 67 89"

Or in uppercase if you prefer:

    <%= @contact.happy_phone(:FR) %>
    #=> "01 23 45 67 89"

Or maybe with a string:

    <%= @contact.happy_phone('fr') %>
    #=> "01 23 45 67 89"

If you want an international format:

    <%= @contact.happy_inter_phone(:fr) %>
    #=> "+33 1 23 45 67 89"

What if you want dots as separator?

    <%= @contact.happy_phone(:fr, '.') %>
    #=> "01.23.45.67.89"

What if your model have several phone attributes, like `phone1`, `phone2`
and `mobile`, or maybe `portable`?

    <%= @contact.happy_phone1(:fr) %>
    #=> "11 11 11 11 11"
    <%= @contact.happy_phone2(:fr) %>
    #=> "22 22 22 22 22"
    <%= @contact.happy_mobile(:fr) %>
    #=> "33 33 33 33 33"
    ...

Now, imagine you live in Belgium, where phone numbers could have 2 formats:

    <%= @contact.happy_phone(:be) %>
    #=> "03 111 22 33"

    <%= @contact.happy_phone2(:be) %>
    #=> "063 11 22 33"

    <%= @contact.happy_inter_phone(:be) %>
    #=> "+32 3 111 22 33"

    <%= @contact.happy_inter_phone2(:be) %>
    #=> "+32 63 11 22 33"

And if *Happy Phone Number* don't know your country, you could use a
**mask formatting**:

    <%= @contact.happy_mask_phone("#### ###-###") %>
    #=> "0123 456-789"


Install
-------------------------

Add this line in your Gemfile:

    gem 'happy_phone_number'


Usage
--------------------------

Add the `happy_phone_number` method in your model. Example:

    class Contact < ActiveRecord::Base
      attr_accessible :email, :name, :phone, :fax
      happy_phone_number
    end

Now, in your views, you can use:

    happy_phone(:fr)
    happy_fax(:fr)
    ...

Supported Countries
--------------------------

<table>
<tr>
  <th>Name</th>
  <th>Symbol</th>
</tr>
<tr>
  <td>Belgium</td>
  <td>:be</td>
</tr>
<tr>
  <td>Denmark</td>
  <td>:dk</td>
</tr>
<tr>
  <td>France</td>
  <td>:fr</td>
</tr>
<tr>
  <td>Iceland</td>
  <td>:is</td>
</tr>
</table>

Dependencies
--------------------------

  * ruby >= 1.9.2
  * rails >= 3.2

Contributing
-------------------------

*I don't know much about format and convention for every possible country.
Feel free to send me documentation for the phone number formatting and
convention for your country. Or contribute to the code by adding any
missing formatter.*

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

### How to add a formatter for your country

Name the file as follow:

    aa_format.rb

where "aa" is the ISO code of the country.
See [ISO country codes] [1].

Put this file in the lib/happy_phone_number/ folder. Then follow the structure of
an existed formatter:

  * fr_format.rb is a good starting point
  * dk_format.rb is the simplest one
  * be_format.rb is a little more complex

Feel free to ask for help if you have any troubles.


### There is many other ways to contibute to FIXME

1. Report any bug
2. Give me ideas
3. Correct my poor english syntax
4. Blog about *Happy Phone Number*
5. Tell *Happy Phone Number* to your friends and colleagues

License
--------------------------

MIT


Questions and/or Comments
--------------------------

Feel free to email [Xavier Nayrac](mailto:xavier.nayrac@gmail.com)
with any questions, or contact me on [twitter](https://twitter.com/lkdjiin).

[1]: http://en.wikipedia.org/wiki/ISO_3166-1 "ISO country codes on wikipedia"


changelog.markdown
version 0.0.3 (2013-05-18)
==========================

* **API break** Add `#happy_mask_* method`.
	Mask formatting have now its own method. So one can call `#happy_phone`
	with a string:

      happy_phone("fr")


version 0.0.2 (2013-05-08)
==========================

* **API break** `happy_phone(:international)` is removed. You can use
  instead `happy_inter_phone(:fr)`
* Add support for Belgium
* Add support for Denmark
* Add support for Iceland
* Make it easy to add a formatter for a new country
* Add a masking method, to display phone numbers from unsupported countries


version 0.0.1 (2013-05-03)
==========================

First version. Know only about french phone numbers.



lib/happy_phone_number.rb
require 'happy_phone_number/formatter'
require 'happy_phone_number/mask'
require 'happy_phone_number/base_format'

# Public: Happy Phone Number provides easy methods to format phone
# numbers from your Rails models. Virtually any country specific
# format could be supported, including national and international
# format. For the not (yet) supported countries, a general method
# using a simple mask is also provided.
#
# Examples
#
#   class Contact < ActiveRecord::Base
#     attr_accessible :email, :name, :phone, :fax
#     happy_phone_number
#   end
#
#   <%= @contact.happy_phone(:fr) %>
#   # => "01 23 45 67 89"
#
#   <%= @contact.happy_phone('fr') %>
#   # => "01 23 45 67 89"
#
#   <%= @contact.happy_inter_phone(:fr) %>
#   # => "+33 1 23 45 67 89"
#
#   <%= @contact.happy_phone(:be) %>
#   # => "03 111 22 33"
#
#   <%= @contact.happy_mask_phone("#### ###-###") %>
#   # => "0123 456-789"
module HappyPhoneNumber
  extend ActiveSupport::Concern

  included do
  end

  module ClassMethods
    def happy_phone_number(options = {})
    end
  end

  # Public: Format a phone number from a model.
  #
  # args[0] - Either a Symbol|String country or a String formatting mask.
  #           A country must be in ISO 3166-1-alpha-2 format,
  #           both lower and upper case are valids.
  # block   - Not used.
  #
  # Signature
  #
  #   happy_<field>(country, separator)
  #   happy_inter_<field>(country, separator)
  #   happy_mask_<field>(mask)
  #
  # field     - A field name of the model that holds a phone number as a
  #             string.
  # country   - Symbol or String in ISO 3166-1-alpha-2 format.
  # separator - A String used to separate groups of number.
  # mask      - A String formatting mask.
  #
  # Returns the String formatted phone number.
  def method_missing(meth, *args, &block)
    if meth.to_s =~ /^happy_inter_(.+)$/
      happy_format(:international, $1, *args)
    elsif meth.to_s =~ /^happy_mask_(.+)$/
      happy_format(:mask, $1, *args)
    elsif meth.to_s =~ /^happy_(.+)$/
      happy_format(:national, $1, *args)
    else
      super
    end
  end

  # Public: Make the happy_* methods respond to respond_to?.
  def respond_to?(meth, include_private = false)
    if meth.to_s =~ /^happy_.*$/
      true
    else
      super
    end
  end

  private

  # Launch the formatting process.
  #
  # type - Either :national, :international or :mask.
  # meth - The String field from the model.
  # args - Argument(s) passed to the formatting method, like a country,
  #        a mask string, the separator.
  #
  # Returns the String formatted phone number.
  def happy_format(type, meth, *args)
    Formatter.new(send(meth.to_sym), type, args[0], args[1]).format
  end

end

ActiveRecord::Base.send :include, HappyPhoneNumber
